### PR TITLE
Add channel checkpoint

### PR DESCRIPTION
**Pull request by sunby**
issue: https://github.com/milvus-io/milvus/issues/17445
Signed-off-by: sunby <bingyi.sun@zilliz.com>

### CODE REVIEW

File: internal/datacoord/server.go [hunk 1]
```diff
@@ -581,6 +582,7 @@ func (s *Server) handleTimetickMessage(ctx context.Context, ttMsg *msgstream.Dat
 	metrics.DataCoordSyncEpoch.WithLabelValues(pChannelName).Set(float64(utcT))
 
 	s.updateSegmentStatistics(ttMsg.GetSegmentsStats())
+	s.updateChannelCheckpoint(ttMsg.GetChannelName(), ttMsg.GetCheckpoint())
```
Comment (xiaofan-luan) [inline internal/datacoord/server.go:585]: so we simply ignore save etcd error, probably need a log here?

File: internal/datanode/segment_replica.go [hunk 2]
```diff
@@ -840,3 +844,131 @@ func (replica *SegmentReplica) listNotFlushedSegmentIDs() []UniqueID {
...
+func (s *segmentsStats) getOrCreateSegmentStats(segmentID UniqueID) *segmentStats {
+	s.Lock()
+	defer s.Unlock()
+	if _, ok := s.segmentsStats[segmentID]; !ok {
+		s.segmentsStats[segmentID] = newSegmentStats()
+	}
+	return s.segmentsStats[segmentID]
+}
+
+func (s *segmentsStats) getLeastStartPosition() *internalpb.MsgPosition {
```
Comment (xiaofan-luan) [inline internal/datanode/segment_replica.go:900]: getEarliestStartPosition

**Comment by xiaofan-luan** [comment]
@congqixia pls also help on reviewing it, thanks

File: internal/datacoord/meta.go [hunk 3]
```diff
@@ -40,13 +41,16 @@ const (
 	channelRemovePrefix = metaPrefix + "/channel-removal"
 
 	removeFlagTomestone = "removed"
+
+	channelCheckpointPrefix = "metaPrefix" + "/checkpoints"
```
[the remark this answers is not in the document]
Comment (sunby) [reply internal/datacoord/meta.go:45]: > is it metaPrefix or "metaPrefix"

yes, it's a typo

File: internal/datanode/segment_replica.go [hunk 4]
```diff
@@ -840,3 +844,131 @@ func (replica *SegmentReplica) listNotFlushedSegmentIDs() []UniqueID {
...
+
+func newChannelConsumeStats() *channelConsumeStats {
+	return &channelConsumeStats{
+		segmentsStats: newSegmentsStats(),
+	}
+}
+
+type segmentsStats struct {
+	sync.RWMutex
+	segmentsStats map[UniqueID]*segmentStats
```
[the remark this answers is not in the document]
Comment (sunby) [reply internal/datanode/segment_replica.go:888]: > when to cleanup remove segmentsStats?

maybe we should clean it every time we get the least pos?

File: internal/datanode/flow_graph_delete_node.go [hunk 5]
```diff
@@ -211,18 +214,32 @@ func (dn *deleteNode) Operate(in []Msg) []Msg {
...
+		}
+	}
+
+	// update segment buffer range
+	startPosition, endPosition := fgMsg.startPositions[0], fgMsg.endPositions[0]
+	for segmentID := range segmentsWithDeletedData {
+		channelStats.update(segmentID, startPosition, endPosition)
 	}
+	// update consume position here because we have to update segment range first to avoid losing data
+	channelStats.setConsumePosition(startPosition)
```
[the remark this answers is not in the document]
Comment (sunby) [reply internal/datanode/flow_graph_delete_node.go:242]: > why update in deleteNode but not insertNode?

deleteNode and insertNode handle data concurrently. If we update in insertNode, larger consume position may be recorded first and less delete position is recorded later.

**Pull request closed without merging**